sns/parity_math.c

```diff
@@ -139,57 +347,95 @@ static void forward_rectification(struct m0_sns_ir *ir,
  * for computing failed parity.
  */
 static void failed_data_blocks_xform(struct m0_sns_ir *ir);
+#endif /* !ISAL_ENCODE_ENABLED */
+
 static inline bool is_valid_block_idx(const  struct m0_sns_ir *ir,
 				      uint32_t block_idx);
 
+#if !ISAL_ENCODE_ENABLED
 static bool is_data(const struct m0_sns_ir *ir, uint32_t index);
-
+#endif /* !ISAL_ENCODE_ENABLED */
 
 static bool is_usable(const struct m0_sns_ir *ir,
 		      const struct m0_bitmap *in_bmap,
 		      struct m0_sns_ir_block *failed_block);
 
 static uint32_t last_usable_block_id(const struct m0_sns_ir *ir,
 				     uint32_t block_idx);
+
+#if !ISAL_ENCODE_ENABLED
 static inline  bool are_failures_mixed(const struct m0_sns_ir *ir);
 
 static inline const struct m0_matrix* recovery_mat_get(const struct m0_sns_ir
 						       *ir,
 						       uint32_t failed_idx);
+#endif /* !ISAL_ENCODE_ENABLED */
+
+static inline uint32_t ir_blocks_count(const struct m0_sns_ir *ir);
 
-static inline uint32_t block_count(const struct m0_sns_ir *ir);
+#if ISAL_ENCODE_ENABLED
+/**
+ * Initialize fields required for incremental recovery using Intel ISA.
+ * @param[in] ir       - Pointer to incremental recovery structure.
+ * @retval    0          Success
+ * @retval    -ENOMEM    Failed to allocate memory
+ */
+static int isal_ir_init(struct m0_sns_ir *ir);
 
-static void (*calculate[M0_PARITY_CAL_ALGO_NR])(struct m0_parity_math *math,
-						const struct m0_buf *data,
-						struct m0_buf *parity) = {
+/**
+ * Free fields initialized for incremental recovery using Intel ISA.
+ * @param[in] ir - pointer to incremental recovery structure.
+ */
+static void isal_ir_fini(struct m0_sns_ir *ir);
+#endif /* ISAL_ENCODE_ENABLED */
+
+static int (*calculate[M0_PARITY_CAL_ALGO_NR])(struct m0_parity_math *math,
+					       const struct m0_buf *data,
+					       struct m0_buf *parity) = {
 	[M0_PARITY_CAL_ALGO_XOR] = xor_calculate,
+#if ISAL_ENCODE_ENABLED
+	[M0_PARITY_CAL_ALGO_ISA] = isal_encode,
+#else
 	[M0_PARITY_CAL_ALGO_REED_SOLOMON] = reed_solomon_encode,
+#endif /* ISAL_ENCODE_ENABLED */
```

Comment:
This is confusing, because ISA is used to calculate Reed-Solomon parity.

Reply:
Sanjog may have more comments.
There are two ways to use ISA:

* Only use ISA math low level math functions. We will be still using reed_solomon_encode() which will use ISA math functions instead of Galois math functions.


* We will use ISA encode/decode functions. This is equivalent to reed_solomon_encode(), but they have different algorithms. So, the same data will generate different parity.  This is the option we are going to use.

Reply:
- As Hua said in point 2, the Intel ISA encoding and recovery APIs are equivalent to reed_solomon_encode() and reed_solomon_recover() function. But the matrix used for calculating Reed-Solomon parity in Intel ISA is different than the current implementation (vandermonde matrix) in Motr code.
- In Intel ISA, currently we are using below matrix
```
Vandermonde matrix example of encoding coefficients where 
high portion of matrix is identity matrix I and 
lower portion is constructed as 2^{i*(j-k+1)} i:{0,k-1} j:{k,m-1}.
where m = number of rows in matrix = data_count + parity_count
and k = number of columns in matrix = data_count.
```
FYI-There is Cauchy matrix support also in ISA.
- Hence the parity calculated using Intel ISA is different compared to the one calculated using reed_solomon_encode().
- So even though it is reed-solomon, we are selecting which implementation to use. i.e. from Intel ISA or from Motr code.

Reply:
Removed `M0_PARITY_CAL_ALGO_ISA`. Used `M0_PARITY_CAL_ALGO_REED_SOLOMON` for Galois as well as Intel ISA implementation. Refer [below comment](https://github.com/Seagate/cortx-motr/pull/667#discussion_r659767012).